File: packages/typescript/src/features/advanced-capabilities/providers/followUps.ts
import { requireProviderSpec } from "../../../generated/spec-helpers.ts";
import type { FollowUpService } from "../../../services/followUp.ts";
import type {
	IAgentRuntime,
	Memory,
	Provider,
	ProviderResult,
	State,
} from "../../../types/index.ts";

// Get text content from centralized specs
const spec = requireProviderSpec("FOLLOW_UPS");

export const followUpsProvider: Provider = {
	name: spec.name,
	description: spec.description,
	get: async (
		runtime: IAgentRuntime,
		_message: Memory,
		_state: State,
	): Promise<ProviderResult> => {
		const followUpService = runtime.getService("follow_up") as FollowUpService;
		if (!followUpService) {
			runtime.logger.warn("[FollowUpsProvider] FollowUpService not available");
			return { text: "", values: {}, data: {} };
		}

		// Get upcoming follow-ups for the next 7 days
		const upcomingFollowUps = await followUpService.getUpcomingFollowUps(
			7,
			true,
		);

		if (upcomingFollowUps.length === 0) {
			return {
				text: "No upcoming follow-ups scheduled.",
				values: { followUpCount: 0 },
				data: {},
			};
		}

		const now = Date.now();
		const contactIds = Array.from(
			new Set(upcomingFollowUps.map((f) => f.contact.entityId)),
		);
		const entities = await Promise.all(
			contactIds.map((id) => runtime.getEntityById(id)),
		);
		const entityNames = new Map<string, string>();
		for (let i = 0; i < contactIds.length; i += 1) {
			entityNames.set(contactIds[i], entities[i]?.names[0] || "Unknown");
		}

		const overdue: typeof upcomingFollowUps = [];
		const upcoming: typeof upcomingFollowUps = [];
		const scheduledAtMs = new Map<string, number>();
		for (const item of upcomingFollowUps) {
			const scheduledAt = item.task.metadata?.scheduledAt
				? new Date(item.task.metadata.scheduledAt as string).getTime()
				: 0;
			if (item.task.id) {
				scheduledAtMs.set(item.task.id, scheduledAt);
			}
			if (scheduledAt < now) {
				overdue.push(item);
			} else {
				upcoming.push(item);
			}
		}

		// Build text summary
		let textSummary = `You have ${upcomingFollowUps.length} follow-up${upcomingFollowUps.length !== 1 ? "s" : ""} scheduled:\n`;

		if (overdue.length > 0) {
			textSummary += `\nOverdue (${overdue.length}):\n`;
			for (const f of overdue) {
				const name = entityNames.get(f.contact.entityId) || "Unknown";
				const scheduledAt = f.task.id ? (scheduledAtMs.get(f.task.id) ?? 0) : 0;

				textSummary += `- ${name}`;
				if (scheduledAt > 0) {
					const daysOverdue = Math.floor(
						(now - scheduledAt) / (1000 * 60 * 60 * 24),
					);
					textSummary += ` (${daysOverdue} day${daysOverdue !== 1 ? "s" : ""} overdue)`;
				}
				if (f.task.metadata?.reason) {
					textSummary += ` - ${f.task.metadata.reason}`;
				}
				textSummary += "\n";
			}
		}

		if (upcoming.length > 0) {
			textSummary += `\nUpcoming (${upcoming.length}):\n`;
			for (const f of upcoming) {
				const name = entityNames.get(f.contact.entityId) || "Unknown";
				const scheduledAt = f.task.id ? (scheduledAtMs.get(f.task.id) ?? 0) : 0;

				textSummary += `- ${name}`;
				if (scheduledAt > 0) {
					const daysUntil = Math.ceil(
						(scheduledAt - now) / (1000 * 60 * 60 * 24),
					);
					if (daysUntil === 0) {
						textSummary += " (today)";
					} else if (daysUntil === 1) {
						textSummary += " (tomorrow)";
					} else {
						textSummary += ` (in ${daysUntil} days)`;
					}
				}
				if (f.task.metadata?.reason) {
					textSummary += ` - ${f.task.metadata.reason}`;
				}
				textSummary += "\n";
			}
		}

		// Get follow-up suggestions
		const suggestions = await followUpService.getFollowUpSuggestions();

		if (suggestions.length > 0) {
			textSummary += `\nSuggested follow-ups:\n`;
			suggestions.slice(0, 3).forEach((s) => {
				textSummary += `- ${s.entityName} (${s.daysSinceLastContact} days since last contact)\n`;
			});
		}

		return {
			text: textSummary.trim(),
			values: {
				followUpCount: upcomingFollowUps.length,
				overdueCount: overdue.length,
				upcomingCount: upcoming.length,
				suggestionsCount: suggestions.length,
			},
			data: {
				followUpCount: upcomingFollowUps.length,
				overdueCount: overdue.length,
				upcomingCount: upcoming.length,
				suggestionsCount: suggestions.length,
			},
		};
	},
};

File: packages/python/elizaos/features/basic_capabilities/providers/follow_ups.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_provider_spec
from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

# Get text content from centralized specs
_spec = require_provider_spec("FOLLOW_UPS")


async def get_follow_ups_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    from elizaos.features.advanced_capabilities.services.follow_up import FollowUpService

    follow_up_service = runtime.get_service("follow_up")
    if not follow_up_service or not isinstance(follow_up_service, FollowUpService):
        return ProviderResult(text="", values={}, data={})

    upcoming = await follow_up_service.get_upcoming_follow_ups(days_ahead=7, include_overdue=True)

    if not upcoming:
        return ProviderResult(
            text="No upcoming follow-ups scheduled.", values={"followUpCount": 0}, data={}
        )

    now = datetime.now(UTC)

    overdue_items: list[dict[str, str]] = []
    upcoming_items: list[dict[str, str]] = []

    for task in upcoming:
        scheduled = datetime.fromisoformat(task.scheduled_at.replace("Z", "+00:00"))
        entity = await runtime.get_entity(str(task.entity_id))
        name = entity.name if entity and entity.name else "Unknown"

        item = {
            "entityId": str(task.entity_id),
            "name": name,
            "scheduledAt": task.scheduled_at,
            "reason": task.reason,
            "priority": task.priority,
        }

        if scheduled < now:
            days_overdue = (now - scheduled).days
            item["daysOverdue"] = str(days_overdue)
            overdue_items.append(item)
        else:
            days_until = (scheduled - now).days
            item["daysUntil"] = str(days_until)
            upcoming_items.append(item)

    text_summary = f"You have {len(upcoming)} follow-up(s) scheduled:\n"

    if overdue_items:
        text_summary += f"\nOverdue ({len(overdue_items)}):\n"
        for item in overdue_items:
            text_summary += f"- {item['name']} ({item['daysOverdue']} days overdue)"
            if item.get("reason"):
                text_summary += f" - {item['reason']}"
            text_summary += "\n"

    if upcoming_items:
        text_summary += f"\nUpcoming ({len(upcoming_items)}):\n"
        for item in upcoming_items:
            days = int(item["daysUntil"])
            if days == 0:
                time_str = "today"
            elif days == 1:
                time_str = "tomorrow"
            else:
                time_str = f"in {days} days"
            text_summary += f"- {item['name']} ({time_str})"
            if item.get("reason"):
                text_summary += f" - {item['reason']}"
            text_summary += "\n"

    suggestions = await follow_up_service.get_follow_up_suggestions()
    if suggestions:
        text_summary += "\nSuggested follow-ups:\n"
        for s in suggestions[:3]:
            text_summary += (
                f"- {s.entity_name} ({s.days_since_last_contact} days since last contact)\n"
            )

    return ProviderResult(
        text=text_summary.strip(),
        values={
            "followUpCount": len(upcoming),
            "overdueCount": len(overdue_items),
            "upcomingCount": len(upcoming_items),
            "suggestionsCount": len(suggestions),
        },
        data={
            "followUpCount": len(upcoming),
            "overdueCount": len(overdue_items),
            "upcomingCount": len(upcoming_items),
            "suggestionsCount": len(suggestions),
        },
    )


follow_ups_provider = Provider(
    name=_spec["name"],
    description=_spec["description"],
    get=get_follow_ups_context,
    dynamic=_spec.get("dynamic", True),
    position=_spec.get("position"),
)
